pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```
@@ -1393,6 +1452,415 @@ public PartitionAssignment getStreamPartitionAssignment(String realtimeTableName
+   * If new partitions are detected, gets them started
+   * If new instances are detected, uses them for new assignments
+   *
+   * So, the method may:
+   * - Add or modify one or more segment metadata znodes
+   * - Update the idealstate, which may fail if some other process updated the idealstate.
+   *
+   * In case idealstate update fails, then we need to start over.
+   * TODO: split this method into multiple smaller methods
+   */
+  @VisibleForTesting
+  protected IdealState validateLLCSegments(final TableConfig tableConfig, IdealState idealState,
+      final int partitionCount)  {
+    final String tableNameWithType = tableConfig.getTableName();
+    final StreamMetadata streamMetadata = new StreamMetadata(tableConfig.getIndexingConfig().getStreamConfigs());
+    if (!idealState.isEnabled()) {
+      LOGGER.info("Skipping validation for disabled table {}", tableNameWithType);
+      return idealState;
+    }
+    final long now = getCurrentTimeMs();
+
+    // Get the metadata for the latest 2 segments of each partition
+    Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> latestMetadata =
+        getLatestMetadata(tableNameWithType);
+
+    // Find partitions for which there is no metadata at all. These are new partitions that we need to start consuming.
+    Set<Integer> newPartitions = new HashSet<>(partitionCount);
+    for (int partition = 0; partition < partitionCount; partition++) {
+      if (!latestMetadata.containsKey(partition)) {
+        LOGGER.info("Found partition {} with no segments", partition);
+        newPartitions.add(partition);
+      }
+    }
+
+    PartitionAssignment partitionAssignment;
+    boolean skipNewPartitions = false;
+    try {
+      partitionAssignment = _partitionAssignmentGenerator.generatePartitionAssignment(tableConfig, partitionCount);
+    } catch (InvalidConfigException e) {
+      _controllerMetrics.addMeteredTableValue(tableNameWithType, ControllerMeter.PARTITION_ASSIGNMENT_GENERATION_ERROR,
```

for new tables, we should actually throw an error here (and not bump any metric).  Perhaps we should pull out the code to create new partition segments...